### DISCUSSION TITLE
Trying to add secrets with pynetbox library

### DISCUSSION THREAD

**Discussion by pmarchal25** [netbox-community/netbox#13409] [General]
Hi everyone,

I have the last version of netbox 3.5.7) installed in my virtual machine (with ubuntu 22.04) and the last version of netbox secrets plugin (1.8.3) and i am trying to migrate the secrets of an older version of netbox (2.8.6). I want to migrate the information using a script in python that uses pynetbox library to interact with the netbox API and i would like to know how could i add a secret from the script (the script will take the users and passwords from an excel and should add the secrets into the plugin of the new netbox version). I have created an user with a ssesion key in the new version and i have the private key of the user which i can add secrets. I have the script done (it should work because i used this script with the older netbox version and it worked), but the problem es that when i try to add the secret appears the following error:

**Traceback (most recent call last):
  File "prueba.py", line 50, in <module>
    r=nb.plugins.secrets.secrets.get(id=1)
  File "/usr/local/lib/python3.6/dist-packages/pynetbox/core/app.py", line 67, in __getattr__
    self._set_session_key()
  File "/usr/local/lib/python3.6/dist-packages/pynetbox/core/app.py", line 78, in _set_session_key
    http_session=self.api.http_session,
  File "/usr/local/lib/python3.6/dist-packages/pynetbox/core/query.py", line 225, in get_session_key
    raise RequestError(req)
pynetbox.core.query.RequestError: The requested url: http://192.168.27.203:8000/api/secrets/get-session-key/?preserve_key=True could not be found.**

I have checked the url and everything is okey (to add secrets i have to use the url api/plugins/secrets/secrets and the code is something really easy. Only we have to create the connection with the api **(nb= pynetbox.api(url=url, token=token, private_key=secret_key)) and after that create the secret: nb.plugins.secrets.secrets.create( and there the secret required fields, like object_id....)**

Can you help me?

Thank you so much.

Best Regards.
Pedro

**Comment by markkuleinio** [comment]
Is this due to the fact that [`pynetbox` 7.0.1](https://github.com/netbox-community/pynetbox/releases/tag/v7.0.1) dropped pre-NetBox 3.3 features (like `secrets` app)? You could try with older `pynetbox`.

**Reply by pmarchal25** [reply]
Hi mark, i don't think so, i am using pynetbox version 6.6.2 (and in netbox 2.8.6 i was using the same pynetbox and i could add secrets in this version). I don't know what is the problem...

**Reply by DanSheps** [reply]
No, he is correct.

We removed secrets from NetBox in [3.0.0](https://docs.netbox.dev/en/stable/release-notes/version-3.0/#breaking-changes), so the functionality is only available through a plugin (I recommend netbox-secrets).

That will 1) Change the API URL to a plugin url which you will have to use accordingly in pynetbox (It should be something like plugins.secrets.secret, it has been awhile since I used pynetbox). 2) Change how you work with secrets in the UI

**Reply by pmarchal25** [reply]
I know it,

I was using another URL (plugins.secrets.secrets…) , but the problem Is that I can’t add secrets due to an issue at the time of reading the private key, I think and I don’t know how to solve it.

**Reply by DanSheps** [reply]
There might be some hardcoded behaviour in the secrets portion of the client.  You would want to open an issue with them.

**Comment by abhi1693** [comment]
You are using an older version of pynetbox with the newer version of secrets and also expecting the code from v2 NetBox to work with v3 NetBox. The approach is wrong in all fronts here.

First of all v1.8.3 is not the latest version. We had released v1.8.5 almost 2 months ago and it the most current version as of now.

Secondly, there has been several changes in the API as `netbox-secrets` is not the original secrets module from the core code. We redesigned it for greater flexibility.

Here is a sample code, beyond this you will have to tailor the code for you own requirements (Note: I have written this code by reading the Secrets API schema and the pynetbox documentation)

```python
import pynetbox
import requests

nb = pynetbox.api(...)
session_key = nb.plugins.secrets.session_keys.create(preserve_key=True, private_key='--- KEY ---')
session = requests.Session()
session.headers = {'X-Session-Key': session_key.session_key}
nb.http_session = session
nb.plugins.secrets.secrets.create(assigned_object_type='dcim.device', assigned_object_id=device.id, plaintext='pass')
```

**Reply by pmarchal25** [reply]
Hi abi,

I am trying to add the secrets with the code that you provieded to me but it still fails in the creation of the session_key. The error is the following:

**Traceback (most recent call last):
  File "prueba.py", line 49, in <module>
    session_key = nb.plugins.secrets.session_keys.create(preserve_key=True, private_key=secret_key)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pynetbox/core/endpoint.py", line 415, in create
    ).post(args[0] if args else kwargs)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pynetbox/core/query.py", line 357, in post
    return self._make_call(verb="post", data=data)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pynetbox/core/query.py", line 247, in _make_call
    raise RequestError(req)
pynetbox.core.query.RequestError: The request failed with code 400 Bad Request but more specific details were not returned in json. Check the NetBox Logs or investigate this exception's error attribute.**

**Comment by benoitpiretatstaffwinbe** [comment]
I had the same kind of problem and I came up with the conclusion that nb.plugins.secrets.session_keys.create() is simply not working so went back to requests directly :

class NetBoxTokenContext:
    def __init__(self, base_url, data, description, cacert, private_key_path=None, passphrase=None):
        self.base_url = base_url
        self.data = data
        self.description = description
        self.cacert = cacert
        self.nb = None
        self.private_key_path = private_key_path
        self.passphrase = passphrase
        self.token_key = None
        self.private_key_path = private_key_path

    def __enter__(self):
        # Create the token and pynetbox API object
        token_data = token_create(self.base_url+"/api/users/tokens/provision/", self.data, self.description)
        if token_data:
            self.token_key = token_data['key']
            session = requests.Session()
            session.verify = self.cacert
            self.nb = pynetbox.api(self.base_url, self.token_key)
            self.nb.http_session = session

        # reate session key
        try:
            with open(self.private_key_path, "r") as key_file:
                private_key_str = key_file.read().strip()
        except FileNotFoundError:
            print(f"Private key file not found: {self.private_key_path}")
            return self.nb # return the pynetbox object with just the token auth.

        if self.private_key_path:
            print(f"API Token: '{token_data['key']}'")  # Debugging line
            session_key = create_session_key_requests(self.base_url, token_data['key'], private_key_str, self.cacert)
            if session_key:
                print(f"Session key created: {session_key}")
                # Update the session with the session key header
                new_session = requests.Session()
                new_session.verify = self.cacert
                new_session.headers.update({"X-Session-Key": session_key})
                self.nb.http_session = new_session # Update the pynetbox session.

            else:
                print("Failed to create session key.")
        return self.nb

    def __exit__(self, exc_type, exc_value, traceback):
        # Delete the token
        if self.nb and self.token_key:
            try:
                tokens = self.nb.users.tokens.all()
                for token in tokens:
                    if hasattr(token, 'description') and token.description == 'NSO-to-NETBOX' :
                        token.delete()
                        ##print(f"Token {self.token_key} deleted.")
                        break
            except Exception as e:
                #self.log.infof"Error deleting token: {e}")
                pass
